### PR TITLE
Support for OAuth in DynamoSandbox

### PR DESCRIPTION
**Pull request by pboyer**
### Objective

This PR paves the way for authentication in Dynamo standalone.  The authentication mechanism has been generalized and new UI has been introduced.  Generally, Dynamo should work and look exactly as it did before, but with hidden powers.  :)

**Sign in:**

signin

**The sign in dialog:**

signingin

**After signing in:**

username

### Related changes

**GregClientNET** : The `IAuthProvider` interface has been extended to more robustly support host-provided & standalone authentication.  The `IGregClient` interface has been created to better support mocking for `PackageManagerClient.`

https://github.com/DynamoDS/GRegClientNET/commit/a33f02701726a4075f264d3acf2e8fa74588acf1

**GregOAuth** : Both Revit and standalone `IAuthProvider`'s have been extended for the new interface and generally cleaned up.

Private repo:

https://github.com/DynamoDS/GRegOAuth/commit/3827dbf5512fb86cac5d01618f4fde279535adcf

**DynamoRevit** : Generally cleaned up and some of the code has been offloaded into the GregOAuth project.  

https://github.com/DynamoDS/DynamoRevit/pull/124

### Reviewers

@lukechurch 

@elayabharath Take a look at the UI.

### CODE REVIEW

File: src/DynamoSandbox/DynamoSandbox.csproj [hunk 1]
```diff
@@ -29,14 +29,27 @@
...
+    <UseVSHostingProcess>true</UseVSHostingProcess>
   </PropertyGroup>
   <ItemGroup>
+    <Reference Include="Greg, Version=1.0.0.0, Culture=neutral, processorArchitecture=MSIL">
+      <SpecificVersion>False</SpecificVersion>
+      <HintPath>..\..\extern\greg\Greg.dll</HintPath>
+    </Reference>
+    <Reference Include="GregStandaloneAuth">
+      <HintPath>..\..\extern\greg\GregStandaloneAuth.dll</HintPath>
+    </Reference>
```
Comment (pboyer) [inline src/DynamoSandbox/DynamoSandbox.csproj]: This DLL is not provided.

File: src/AssemblySharedInfoGenerator/AssemblyInfoGenerator.csproj [hunk 2]
```diff
@@ -26,6 +26,7 @@
     <DefineConstants>DEBUG;TRACE</DefineConstants>
     <ErrorReport>prompt</ErrorReport>
     <WarningLevel>4</WarningLevel>
+    <UseVSHostingProcess>true</UseVSHostingProcess>
```
Comment (pboyer) [inline src/AssemblySharedInfoGenerator/AssemblyInfoGenerator.csproj]: Need to get this fluff out before submitting.

File: src/DynamoCore/PackageManager/PackageManagerClient.cs [hunk 3]
```diff
@@ -286,44 +177,15 @@ private PackageUploadHandle PublishPackage(bool isNewVersion, Package l, List<st
             });
 
             return packageUploadHandle;
-
         }
 
-        public class PackageManagerResult
```
Comment (pboyer) [inline src/DynamoCore/PackageManager/PackageManagerClient.cs]: Moved to separate file.

File: src/DynamoCoreWpf/NodeViewCustomization/NodeViewCustomizations/DSVarArgFunctionNodeViewCustomization.cs [hunk 4]
```diff
@@ -2,15 +2,16 @@
...
-using Dynamo.Nodes;
 
 namespace Dynamo.Wpf
 {
-    internal class DSVarArgFunctionNodeViewCustomization : INodeViewCustomization<Dynamo.Nodes.DSVarArgFunction>
+    internal class DSVarArgFunctionNodeViewCustomization : INodeViewCustomization<DSVarArgFunction>
     {
-        public void CustomizeView(Nodes.DSVarArgFunction nodeModel, Controls.NodeView nodeView)
+        public void CustomizeView(Nodes.DSVarArgFunction nodeModel, NodeView nodeView)
         {
```
Comment (pboyer) [inline src/DynamoCoreWpf/NodeViewCustomization/NodeViewCustomizations/DSVarArgFunctionNodeViewCustomization.cs]: Minor cleanup.

File: src/DynamoCoreWpf/Properties/AssemblyInfo.cs [hunk 5]
```diff
@@ -37,3 +37,4 @@
 [assembly: InternalsVisibleTo("DynamoCoreUITests")]
 [assembly: InternalsVisibleTo("DynamoMSOfficeTests")]
 [assembly: InternalsVisibleTo("CoreNodeModelsWpf")]
+[assembly: InternalsVisibleTo("DynamoSandbox")]
```
Comment (pboyer) [inline src/DynamoCoreWpf/Properties/AssemblyInfo.cs]: Required for `LoginService`

File: src/DynamoCoreWpf/ViewModels/PackageManager/PackageManagerClientViewModel.cs [hunk 6]
```diff
@@ -186,13 +221,6 @@ public List<PackageManagerSearchElement> ListAll()
             return CachedPackageList;
         }
 
-        public List<PackageManagerSearchElement> Search(string search, int maxNumSearchResults)
-        {
-            return Model.Search(search, maxNumSearchResults)
-                               .Select((header) => new PackageManagerSearchElement(Model, header))
-                               .ToList();
-        }
-
```
Comment (pboyer) [inline src/DynamoCoreWpf/ViewModels/PackageManager/PackageManagerClientViewModel.cs]: Remote package search.  This isn't in use.  Removing it here to prevent confusion.

File: src/DynamoCoreWpf/ViewModels/PackageManager/PackageManagerSearchViewModel.cs [hunk 7]
```diff
@@ -700,36 +700,6 @@ internal IEnumerable<PackageManagerSearchElementViewModel> Search(string query)
...
-                PackageManagerClientViewModel.Search(search, MaxNumSearchResults)
-                    .Select(x => new PackageManagerSearchElementViewModel(x)).ToList();
-
-            if (emptySearch)
-            {
-                Sort(results, this.SortingKey);
-            }
-
-            return results;
-        }
```
Comment (pboyer) [inline src/DynamoCoreWpf/ViewModels/PackageManager/PackageManagerSearchViewModel.cs]: As noted above, removed.

File: src/DynamoCoreWpf/Views/Core/DynamoView.xaml.cs [hunk 8]
```diff
@@ -175,6 +175,15 @@ void DynamoView_SizeChanged(object sender, SizeChangedEventArgs e)
...
         }
 
+        void InitializeLogin()
+        {
+            if ( dynamoViewModel.ShowLogin && dynamoViewModel.PackageManagerClientViewModel.HasAuthProvider)
+            {
+                var login = new Login(dynamoViewModel.PackageManagerClientViewModel);
+                loginGrid.Children.Add(login);
+            }
+        }
```
Comment (pboyer) [inline src/DynamoCoreWpf/Views/Core/DynamoView.xaml.cs]: Add the login dialog to the view

File: src/DynamoSandbox/Program.cs [hunk 9]
```diff
@@ -1,44 +1,56 @@
...
-                    DynamoModel = model
+                    DynamoModel = model,
+                    ShowLogin = true
                 });
 
             var view = new DynamoView(viewModel);
 
+            var loginService = new LoginService(view, new DispatcherSynchronizationContext(Dispatcher.CurrentDispatcher));
+            authProvider.RequestLogin += loginService.ShowLogin;
+
```
Comment (pboyer) [inline src/DynamoSandbox/Program.cs]: Will be commented out.

File: src/DynamoSandbox/app.config [hunk 10]
```diff
@@ -8,4 +8,7 @@
       <probing privatePath="dll"/>
     </assemblyBinding>
   </runtime>
+  <appSettings>
+    <add key="authAddress" value="https://accounts-staging.autodesk.com/" />
+  </appSettings>
```
Comment (pboyer) [inline src/DynamoSandbox/app.config]: The end point for accounts on the package manager staging server.

File: src/DynamoCore/App.config [hunk 11]
```diff
@@ -0,0 +1,6 @@
+﻿<?xml version="1.0" encoding="utf-8" ?>
+<configuration>
+  <appSettings>
+    <add key="packageManagerAddress" value="http://107.20.146.184" />
```
Comment (pboyer) [inline src/DynamoCore/App.config]: The package manager staging server.  Will maintain this for testing with @jnealb.

File: src/DynamoCore/DynamoCore.csproj [hunk 12]
```diff
@@ -94,12 +95,10 @@ limitations under the License.
       <HintPath>..\..\extern\greg\RestSharp.dll</HintPath>
     </Reference>
     <Reference Include="System" />
+    <Reference Include="System.configuration" />
     <Reference Include="System.Core">
       <RequiredTargetFramework>3.5</RequiredTargetFramework>
     </Reference>
-    <Reference Include="System.Reactive.Interfaces">
-      <HintPath>..\packages\Rx-Interfaces.2.2.5\lib\net40\System.Reactive.Interfaces.dll</HintPath>
-    </Reference>
```
Comment (pboyer) [inline src/DynamoCore/DynamoCore.csproj]: Not sure why this is still in here.

File: src/DynamoCore/PackageManager/PackageManagerClient.cs [hunk 13]
```diff
@@ -1,251 +1,141 @@
...
-        //    if (CachedPackageList.Any(x => x.Id == packageId && x.Maintainers.Contains(un)))
-        //    {
-        //        return true;
-        //    }
-
-        //    var l = ListAll();
-        //    return l.Any(x => x.Id == packageId && x.Maintainers.Contains(un));
-
-        //}
-
```
Comment (pboyer) [inline src/DynamoCore/PackageManager/PackageManagerClient.cs]: I despise having commented out code around.  Will revive when it makes sense.

**Comment by lukechurch** [comment]
@pboyer Sorry, I hadn't seen how sensitive to rotting this was

File: src/DynamoCore/App.config (hunk 11 above)
Comment (lukechurch) [inline src/DynamoCore/App.config]: I'm not entirely happy about having the primary server on HTTPS and the staging on HTTP. OK for now, but lets get this switched over to HTTPS when we can.

File: src/DynamoCore/DynamoCore.csproj [hunk 14]
```diff
@@ -94,12 +95,10 @@ limitations under the License.
       <HintPath>..\..\extern\greg\RestSharp.dll</HintPath>
     </Reference>
     <Reference Include="System" />
+    <Reference Include="System.configuration" />
```
Comment (lukechurch) [inline src/DynamoCore/DynamoCore.csproj]: lowercase 'c'?

File: src/DynamoCore/Models/DynamoModel.cs [hunk 15]
```diff
@@ -16,6 +15,7 @@
 
 using DynamoUnits;
 using DynamoUtilities;
+using Greg;
```
Comment (lukechurch) [inline src/DynamoCore/Models/DynamoModel.cs]: Greg is a really unexplanatory name. As this is an introduction into the core part of the application, please can you put a line comment that states what this is

using Greg; // The Auth provider for the package manager

Alternatively we could alias the namespace?

File: src/DynamoCore/Models/DynamoModel.cs [hunk 16]
```diff
@@ -351,6 +351,8 @@ public struct StartConfiguration
             public bool StartInTestMode { get; set; }
             public IUpdateManager UpdateManager { get; set; }
             public ISchedulerThread SchedulerThread { get; set; }
+            public IAuthProvider AuthProvider { get; set; }
+            public string PackageManagerAddress { get; set; }
```
Comment (lukechurch) [inline src/DynamoCore/Models/DynamoModel.cs]: Does this need setting more than once? Can we make it readonly?

File: src/DynamoCore/Models/DynamoModel.cs [hunk 17]
```diff
@@ -384,15 +386,15 @@ public static DynamoModel Start(StartConfiguration configuration)
             return new DynamoModel(configuration);
         }
 
-        protected DynamoModel(StartConfiguration configuration)
+        protected DynamoModel(StartConfiguration config)
         {
             ClipBoard = new ObservableCollection<ModelBase>();
             MaxTesselationDivisions = 128;
```
Comment (lukechurch) [inline src/DynamoCore/Models/DynamoModel.cs]: This upsets me.
Please can we move this into a constant?

File: src/DynamoCore/Models/DynamoModel.cs [hunk 18]
```diff
@@ -458,7 +460,13 @@ protected DynamoModel(StartConfiguration configuration)
             Logger.Log(
                 string.Format("Dynamo -- Build {0}", Assembly.GetExecutingAssembly().GetName().Version));
 
-            PackageManagerClient = new PackageManagerClient(
+            var url = config.PackageManagerAddress ??
+                      AssemblyConfiguration.Instance.GetAppSetting("packageManagerAddress");
```
Comment (lukechurch) [inline src/DynamoCore/Models/DynamoModel.cs]: This should be in a final const

File: src/DynamoCore/PackageManager/PackageManagerClient.cs [hunk 19]
```diff
@@ -1,251 +1,141 @@
...
-            }
-        }
-
-        #endregion
-
         #region Properties/Fields
 
-        private readonly string rootPkgDir;
-        private readonly CustomNodeManager customNodeManager;
+        [Obsolete] internal static readonly string PackageContainsBinariesConstant =
```
Comment (lukechurch) [inline src/DynamoCore/PackageManager/PackageManagerClient.cs]: Can we add a task to purge these please?

File: src/DynamoCore/PackageManager/PackageManagerClient.cs [hunk 20]
```diff
@@ -1,251 +1,141 @@
...
+            "|ContainsBinaries(5C698212-A139-4DDD-8657-1BF892C79821)";
 
-        [Obsolete]
-        internal readonly static string PackageContainsBinariesConstant = "|ContainsBinaries(5C698212-A139-4DDD-8657-1BF892C79821)";
+        [Obsolete] internal static readonly string PackageContainsPythonScriptsConstant =
+            "|ContainsPythonScripts(58B25C0B-CBBE-4DDC-AC39-ECBEB8B55B10)";
 
-        [Obsolete]
-        internal readonly static string PackageContainsPythonScriptsConstant = "|ContainsPythonScripts(58B25C0B-CBBE-4DDC-AC39-ECBEB8B55B10)";
+        private readonly IGregClient _client;
```
Comment (lukechurch) [inline src/DynamoCore/PackageManager/PackageManagerClient.cs]: hard to tell from the diff if these are really new. If they are they should be in .net naming scheme, i.e. no '_'.

File: src/DynamoCore/PackageManager/PackageManagerClient.cs [hunk 21]
```diff
@@ -1,251 +1,141 @@
...
+        public event Action<LoginState> LoginStateChanged;
 
-        public bool HasAuthenticator
+        /// <summary>
+        ///     Specifies whether the user is logged in or not.
+        /// </summary>
+        public LoginState LoginState
         {
-            get { return this.RequestAuthentication != null; }
+            get { return HasAuthProvider ? _authProvider.LoginState : Greg.AuthProviders.LoginState.LoggedOut; }
```
Comment (lukechurch) [inline src/DynamoCore/PackageManager/PackageManagerClient.cs]: Line too long, pls break

File: src/DynamoCore/PackageManager/PackageManagerClient.cs [hunk 22]
```diff
@@ -1,251 +1,141 @@
...
-            }
+            get { return _authProvider != null; }
         }
 
         #endregion
 
-        private static readonly string serverUrl = "https://www.dynamopackages.com/";
-        
-        public PackageManagerClient(string rootPkgDir,  CustomNodeManager customNodeManager)
+        internal PackageManagerClient(IGregClient client, string rootPkgDir, CustomNodeManager customNodeManager)
```
Comment (lukechurch) [inline src/DynamoCore/PackageManager/PackageManagerClient.cs]: How many implementations of IGregClient are there?

File: src/DynamoCore/PackageManager/PackageManagerClient.cs [hunk 23]
```diff
@@ -1,251 +1,141 @@
...
         {
-            this.rootPkgDir = rootPkgDir;
-            this.customNodeManager = customNodeManager;
+            _rootPkgDir = rootPkgDir;
```
Comment (lukechurch) [inline src/DynamoCore/PackageManager/PackageManagerClient.cs]: more _ urg.

File: src/DynamoCore/PackageManager/PackageManagerClient.cs [hunk 24]
```diff
@@ -1,251 +1,141 @@
...
-            }
-            catch
-            {
-                return false;
+                LoginStateChanged(status);
             }
         }
 
-        public bool Downvote(string packageId)
+        private T TryExecute<T>(Func<T> func, T failureResult)
```
Comment (lukechurch) [inline src/DynamoCore/PackageManager/PackageManagerClient.cs]: This is a support method that really shouldn't be a private method of a class

File: src/DynamoCore/PackageManager/PackageManagerClient.cs [hunk 25]
```diff
@@ -1,251 +1,141 @@
...
-                throw new AuthenticationException(
-                    "It looks like you're not logged into Autodesk 360.  Log in to submit a package.");
-            }
+                var nv = HeaderCollectionDownload.ByEngine("dynamo");
+                var pkgResponse = _client.ExecuteAndDeserializeWithContent<List<PackageHeader>>(nv);
+                return pkgResponse.content;
+            }, new List<PackageHeader>());
+        }
 
+        internal PackageUploadHandle Publish(Package l, List<string> files, bool isNewVersion, bool isTestMode)
```
Comment (lukechurch) [inline src/DynamoCore/PackageManager/PackageManagerClient.cs]: Package 'l'?

Comment (lukechurch) [inline src/DynamoCore/PackageManager/PackageManagerClient.cs]: OK, i"m guessing this is to be similar to an existing legacy API?

File: src/DynamoCore/PackageManager/PackageManagerClient.cs [hunk 26]
```diff
@@ -254,15 +144,17 @@ private PackageUploadHandle PublishPackage(bool isNewVersion, Package l, List<st
...
                     {
-                        var pkg = PackageUploadBuilder.NewPackage(rootPkgDir, customNodeManager, l, files, packageUploadHandle, isTestMode);
+                        var pkg = PackageUploadBuilder.NewPackage(_rootPkgDir, _customNodeManager, l, files,
+                            packageUploadHandle, isTestMode);
                         packageUploadHandle.UploadState = PackageUploadHandle.State.Uploading;
-                        ret = Client.ExecuteAndDeserialize(pkg);
+                        ret = _client.ExecuteAndDeserialize(pkg);
                     }
                     if (ret == null)
                     {
```
Comment (lukechurch) [inline src/DynamoCore/PackageManager/PackageManagerClient.cs]: Localisation of error message

File: src/DynamoCore/Utilities/AssemblyConfiguration.cs [hunk 27]
```diff
@@ -0,0 +1,55 @@
...
+using System.Text;
+
+namespace Dynamo.Utilities
+{
+    /// <summary>
+    ///     A tool for obtaining configuration values from the current Library.  
+    /// </summary>
+    internal sealed class AssemblyConfiguration
+    {
+        private static Configuration _config;
```
Comment (lukechurch) [inline src/DynamoCore/Utilities/AssemblyConfiguration.cs]: _

File: src/DynamoCoreWpf/NodeViewCustomization/NodeViewCustomizations/DSVarArgFunctionNodeViewCustomization.cs (hunk 4 above)
Comment (lukechurch) [inline src/DynamoCoreWpf/NodeViewCustomization/NodeViewCustomizations/DSVarArgFunctionNodeViewCustomization.cs]: +1

File: src/DynamoCoreWpf/Services/LoginService.cs [hunk 28]
```diff
@@ -0,0 +1,60 @@
...
+        {
+            this._parent = parent;
+            this._context = context;
+        }
+
+        public bool ShowLogin(object o)
+        {
+            var url = o as Uri;
+
+            if (o == null) throw new Exception("Invalid URL for login page!");
```
Comment (lukechurch) [inline src/DynamoCoreWpf/Services/LoginService.cs]: InvalidArgumentException

File: src/DynamoCoreWpf/Services/LoginService.cs [hunk 29]
```diff
@@ -0,0 +1,60 @@
...
+
+            var navigateSuccess = false;
+
+            // show the login
+            _context.Send((_) => {
+
+                var window = new BrowserWindow(url)
+                {
+                    Title = "Autodesk Sign In",
```
Comment (lukechurch) [inline src/DynamoCoreWpf/Services/LoginService.cs]: Does this need localisation?

File: src/DynamoCoreWpf/ViewModels/PackageManager/PackageManagerClientViewModel.cs [hunk 30]
```diff
@@ -42,13 +46,46 @@ public ObservableCollection<PackageDownloadHandle> Downloads
...
+            {
+                this.Model.Logout();
+            }
+            else
+            {
+                this.Model.Login();
+            }
+        }
+
+        private bool CanLoginLogout()
```
Comment (lukechurch) [inline src/DynamoCoreWpf/ViewModels/PackageManager/PackageManagerClientViewModel.cs]: Property

File: src/DynamoCoreWpf/ViewModels/PackageManager/PackageManagerClientViewModel.cs [hunk 31]
```diff
@@ -42,13 +46,46 @@ public ObservableCollection<PackageDownloadHandle> Downloads
...
             this.DynamoViewModel = dynamoViewModel;
             Model = model;
             CachedPackageList = new List<PackageManagerSearchElement>();
+
+            this.LoginLogoutCommand = new DelegateCommand(LoginLogout, CanLoginLogout);
+
+            model.LoginStateChanged += b => RaisePropertyChanged("LoginState");
+        }
+
+        private void LoginLogout()
```
Comment (lukechurch) [inline src/DynamoCoreWpf/ViewModels/PackageManager/PackageManagerClientViewModel.cs]: How about "SwitchLoginState()"?

File: src/DynamoSandbox/app.config (hunk 10 above)
Comment (lukechurch) [inline src/DynamoSandbox/app.config]: Can we put a  comment in this file saying this?

File: src/DynamoCore/Models/DynamoModel.cs (hunk 17 above)
Comment (lukechurch) [inline src/DynamoCore/Models/DynamoModel.cs]: We can do that in another PR if needs be

**Comment by lukechurch** [comment]
Lets fix these surface issues and then I'll take another look. Thanks @pboyer

File: src/DynamoCore/App.config (hunk 11 above)
Comment (pboyer) [inline src/DynamoCore/App.config]: Yup.

File: src/DynamoCore/DynamoCore.csproj (hunk 14 above)
Comment (pboyer) [inline src/DynamoCore/DynamoCore.csproj]: How odd

File: src/DynamoCore/Models/DynamoModel.cs (hunk 15 above)
Comment (pboyer) [inline src/DynamoCore/Models/DynamoModel.cs]: Sure and http://adsk-oss.myjetbrains.com/youtrack/issue/MAGN-6087

File: src/DynamoCore/Models/DynamoModel.cs (hunk 16 above)
Comment (pboyer) [inline src/DynamoCore/Models/DynamoModel.cs]: This is passed in the `StartConfiguration` object and then set to a `readonly` property in `PackageManagerClient`.

File: src/DynamoCore/Models/DynamoModel.cs (hunk 17 above)
Comment (pboyer) [inline src/DynamoCore/Models/DynamoModel.cs]: It shouldn't be a constant.  It controls at runtime how precise the tessellation is.  Spelling error, though.

File: src/DynamoCore/PackageManager/PackageManagerClient.cs (hunk 19 above)
Comment (pboyer) [inline src/DynamoCore/PackageManager/PackageManagerClient.cs]: Early on, certain packages used them.  That's they they're still around at all.

File: src/DynamoCore/PackageManager/PackageManagerClient.cs (hunk 20 above)
Comment (pboyer) [inline src/DynamoCore/PackageManager/PackageManagerClient.cs]: Goddam, who actually uses `this.privateMember` though?

File: src/DynamoCore/PackageManager/PackageManagerClient.cs (hunk 22 above)
Comment (pboyer) [inline src/DynamoCore/PackageManager/PackageManagerClient.cs]: Today, 2.

File: src/DynamoCoreWpf/ViewModels/PackageManager/PackageManagerClientViewModel.cs (hunk 30 above)
Comment (pboyer) [inline src/DynamoCoreWpf/ViewModels/PackageManager/PackageManagerClientViewModel.cs]: This is being consumed by a DelegateCommand, can't be a prop.

**Comment by pboyer** [comment]
Ping @lukechurch

File: src/DynamoCore/Models/DynamoModel.cs (hunk 17 above)
Comment (lukechurch) [inline src/DynamoCore/Models/DynamoModel.cs]: Sorry to clarify, the 128 should be a declared constant e.g. 
final int MAX_TESSELATIONS_DIVISIONS_DEFAULT = 128;

File: src/DynamoCore/Models/DynamoModel.cs (hunk 16 above)
Comment (lukechurch) [inline src/DynamoCore/Models/DynamoModel.cs]: OK - for something like this, a class with readonly backing store fiends is generally preferable. readonly in C# lets them be set once, so it doesn't stop the use of initialisers. I don't know off the top of my head whether they play nicely with auto-props.

However it's not a big deal in this case.

File: src/DynamoCoreWpf/ViewModels/PackageManager/PackageManagerClientViewModel.cs [hunk 32]
```diff
@@ -277,7 +302,7 @@ public void ClearCompletedDownloads()
 
         internal void GoToWebsite()
         {
-            Process.Start(Model.Client.BaseUrl);
+            Process.Start(Model.BaseUrl);
```
Comment (lukechurch) [inline src/DynamoCoreWpf/ViewModels/PackageManager/PackageManagerClientViewModel.cs]: Please harden, assert starts with 'http://' or 'https://'

File: src/DynamoSandbox/Program.cs [hunk 33]
```diff
@@ -53,11 +62,11 @@ public static void Main(string[] args)
                 // Running Dynamo sandbox with a command file:
                 // DynamoSandbox.exe /c "C:\file path\file.xml"
                 // 
-                string commandFilePath = string.Empty;
-                for (int i = 0; i < args.Length; ++i)
+                var commandFilePath = string.Empty;
+                for (var i = 0; i < args.Length; ++i)
```
Comment (lukechurch) [inline src/DynamoSandbox/Program.cs]: foreach

File: src/DynamoCoreWpf/Services/LoginService.cs [hunk 34]
```diff
@@ -0,0 +1,54 @@
...
+        {
+            this.parent = parent;
+            this.context = context;
+        }
+
+        public bool ShowLogin(object o)
+        {
+            var url = o as Uri;
+
+            if (o == null) throw new ArgumentException("Invalid URL for login page!");
```
Comment (lukechurch) [inline src/DynamoCoreWpf/Services/LoginService.cs]: Localise

**Comment by lukechurch** [comment]
With additional fixes, LGTM!

File: src/DynamoCore/Models/DynamoModel.cs (hunk 16 above)
Comment (pboyer) [inline src/DynamoCore/Models/DynamoModel.cs]: Object initializers are just syntactic sugar on assignments after the object is constructed.  You can't use them to assign `readonly` properties.

File: src/DynamoSandbox/Program.cs (hunk 33 above)
Comment (pboyer) [inline src/DynamoSandbox/Program.cs]: Nah, that doesn't make sense.
